src/js/lib/search.js

```diff
@@ -0,0 +1,344 @@
+
+const OBJECT_TYPE = "type"
+
+const QUERY_FIELDS = [
+  "title.english^3",
+  "short_description.english^2",
+  "full_description.english",
+  "topics",
+  "platform",
+  "course_id",
+  "coursenum^5",
+  "offered_by"
+]
+
+export const RESOURCE_QUERY_NESTED_FIELDS = [
+  "runs.year",
+  "runs.semester",
+  "runs.level",
+  "runs.instructors^5"
+]
+
+export const LR_TYPE_ALL = [
+  LR_TYPE_COURSE,
+  LR_TYPE_VIDEO,
+  LR_TYPE_PODCAST,
+  LR_TYPE_PODCAST_EPISODE,
+  LR_TYPE_RESOURCEFILE
+]
+
+const searchFields = type => {
+  if (type === LR_TYPE_COURSE) {
+    return COURSE_QUERY_FIELDS
+  } else if (type === LR_TYPE_VIDEO) {
+    return VIDEO_QUERY_FIELDS
+  } else if (type === LR_TYPE_PODCAST) {
+    return PODCAST_QUERY_FIELDS
+  } else if (type === LR_TYPE_PODCAST_EPISODE) {
+    return PODCAST_EPISODE_QUERY_FIELDS
+  } else if (type === LR_TYPE_RESOURCEFILE) {
+    return ["content", "title", "short_description"]
+  } else {
+    return LIST_QUERY_FIELDS
+  }
+}
+
+const PODCAST_QUERY_FIELDS = [
+  "title.english^3",
+  "short_description.english^2",
+  "full_description.english",
+  "topics"
+]
+
+const PODCAST_EPISODE_QUERY_FIELDS = [
+  "title.english^3",
+  "short_description.english^2",
+  "full_description.english",
+  "topics",
+  "series_title^2"
+]
+
+const COURSE_QUERY_FIELDS = [
+  "title.english^3",
+  "short_description.english^2",
+  "full_description.english",
+  "topics",
+  "platform",
+  "course_id",
+  "coursenum^5",
+  "offered_by"
+]
+const VIDEO_QUERY_FIELDS = [
+  "title.english^3",
+  "short_description.english^2",
+  "full_description.english",
+  "transcript.english^2",
+  "topics",
+  "platform",
+  "video_id",
+  "offered_by"
+]
+
+const LIST_QUERY_FIELDS = [
+  "title.english^3",
+  "short_description.english^2",
+  "topics"
+]
+
+export const isDoubleQuoted = string => /^".+"$/.test(string)
+
+export const buildSearchQuery = ({ text, from, size, sort }) => {
+  let builder = bodybuilder()
+
+  if (from !== undefined) {
+    builder = builder.from(from)
+  }
+  if (size !== undefined) {
+    builder = builder.size(size)
+  }
+  if (sort !== undefined) {
+    const { field, option } = sort
+    builder.sort(field, option)
+  }
+
+  const types = [LR_TYPE_COURSE, LR_TYPE_VIDEO]
+
+  for (const type of LR_TYPE_ALL) {
+    const queryType = isDoubleQuoted(text) ? "query_string" : "multi_match"
+    const textQuery = emptyOrNil(text) ?
+      {} :
+      {
+        should: [
+          {
+            [queryType]: {
+              query:  text,
+              fields: searchFields(type)
+            }
+          },
+          type === LR_TYPE_COURSE ?
+            {
+              nested: {
+                path:  "runs",
+                query: {
+                  [queryType]: {
+                    query:  text,
+                    fields: RESOURCE_QUERY_NESTED_FIELDS
+                  }
+                }
+              }
+            } :
+            null,
+          type === LR_TYPE_COURSE ?
```

Comment:
You. can probably combine this conditional with the one above (both `type === LR_TYPE_COURSE ?`)